packages/middleware/rate-limiter/src/sliding-window.ts

```diff
@@ -0,0 +1,125 @@
+import { Rate, RateLiteral, parseRate } from './rate'
+import { Slot, SlotProvider } from './slot'
+import { types } from '@mondrian-framework/model'
+import { functions } from '@mondrian-framework/module'
+
+//inspired by https://blog.cloudflare.com/counting-things-a-lot-of-different-things/
+
+/**
+ * Rate limit at `rateLimit` in the `samplingPeriod`.
+ * Example of maximum 10 request every 2 minutes:
+ * ```typescript
+ * new SlidingWindow({ rateLimit: 10, samplingPeriod: "2 minutes" })
+ * ```
+ */
+export class SlidingWindowProvider {
+  private readonly slidingWindows: Map<string, SlidingWindow> = new Map()
+  private readonly rate: Rate
+  private readonly slotProvider: SlotProvider
+
+  constructor({ rate, slotProvider }: { rate: Rate | RateLiteral; slotProvider: SlotProvider }) {
+    this.rate = typeof rate === 'string' ? parseRate(rate) : rate
+    this.slotProvider = slotProvider
+  }
+
+  get({ key }: { key: string }): SlidingWindow {
```

Comment:
Why is it taking a single field object and not just a string argument? What's the reasoning behind this?